lib/signalfx/signalflow/websocket.rb

```diff
@@ -0,0 +1,303 @@
+        channel.computation = computation
+        computation.channel = channel
+        lock.synchronize do
+          job_started_cv.signal
+        end
+      end
+    end
+
+    yield channel.name
+
+    lock.synchronize do
+      if computation.nil?
+        timeout = 20
+        job_started_cv.wait(lock, timeout)
+        if computation.nil?
+          raise "Computation did not start within #{timeout} seconds"
+        end
+      end
+    end
+
+    computation
+  end
+
+  def execute(program, start: nil, stop: nil, resolution: nil, max_delay: nil, persistent: nil)
+    start_job do |channel_name|
+      send_msg({
+        :type => "execute",
+        :channel => channel_name,
+        :program => program,
+        :start => start,
+        :stop => stop,
+        :resolution => resolution,
+        :max_delay => max_delay,
+        :persistent => persistent,
+        :compress => @compress,
+      }.reject!{|k,v| v.nil?}.to_json)
+    end
+  end
+
+  def preflight(program, start, stop, resolution: nil, max_delay: nil)
+    start_job do |channel_name|
+      send_msg({
+        :type => "preflight",
+        :channel => channel_name,
+        :program => program,
+        :start => start,
+        :stop => stop,
+        :resolution => resolution,
+        :max_delay => max_delay,
+        :compress => @compress,
+      }.reject!{|k,v| v.nil?}.to_json)
+    end
+  end
+
+  def start(program, start: nil, stop: nil, resolution: nil, max_delay: nil)
+    send_msg({
+      :type => "start",
+      :program => program,
+      :start => start,
+      :stop => stop,
+      :resolution => resolution,
+      :max_delay => max_delay,
+    }.reject!{|k,v| v.nil?}.to_json)
+  end
+
+  def stop(handle, reason)
+    send_msg({
+      :type => "stop",
+      :handle => handle,
+      :reason => reason,
+    }.reject!{|k,v| v.nil?}.to_json)
+  end
+
+  # This doesn't actually work on the backend yet
+  def attach(handle, filters: nil, resolution: nil)
+    channel = make_new_channel
+
+    send_msg({
+      :type => "attach",
+      :channel => channel.name,
+      :handle => handle,
+      :filters => filters,
+      :resolution => resolution,
+      :compress => @compress,
+    }.reject!{|k,v| v.nil?}.to_json)
+
+    channel
+  end
+
+  def detach(channel, reason=nil)
+    send_msg({
+      :type => "detach",
+      :channel => channel,
+      :reason => reason,
+    }.to_json)
+
+    # There is no response message from the server signifying detach complete
+    # and there could be messages coming in even after the detach request is
+    # sent.  Therefore, use a sentinal value in place of the callback block so
+    # that the message receiver logic can distinguish this case from some
+    # anomolous case (say, due to bad logic in the code).
+    @chan_callbacks[channel] = DETACHED
+  end
+
+  def close
+    if @ws
+      @ws.close
+    end
+  end
+
+  def send_msg(msg)
+    @lock.synchronize do
```

Comment:
Why do you need to synchronize? I don't think you need to make any thread safety guarantees at this layer.

Reply:
Since it lazily connected upon the first message sent, if you were using the same client to start two computations in parallel with a fresh client, only one should actually create the websocket connection that both will share.  Without locks, the connection could be made twice and it would probably cause auth logic to get messed up too.

The send itself is included in the lock because I'm not very confident about the thread-safety of the websocket client lib.